### PR TITLE
Auto-Linting: Spelling Black Style Checker

### LINKED ISSUE #329: Make the entire code base PEP8 compliant.

### PR DESCRIPTION
**Pull request by mrmundt**
## Summary/Motivation:
Finally, this is the PR that adds in the auto-linting pre-req for the regular testing suite.

## Changes proposed in this PR:
- Spell checker (with correctly spelled but falsely identified words represented in `.github/workflows/typos.toml`
- Black style checker (current version: 23.3.0 - which adds some extra spaces between classes/functions. Ignores the `examples/pyomobook/python-ch/BadIndent.py` file as it is intentionally wrong)
- Introduces the `.git-blame-ignore-revs` file which can be used to ignore all of the black style changes when running a `git blame` command:
   1. `git blame --ignore-revs-file .git-blame-ignore-revs`
   2. `git config blame.ignoreRevsFile .git-blame-ignore-revs`

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/core/base/range.py [hunk 1]
```diff
@@ -611,7 +611,7 @@ def range_difference(self, other_ranges):
     def range_intersection(self, other_ranges):
         """Return the intersection between this range and a set of other ranges.
 
-        Paramters
+        Parameters
         ---------
```
Comment (jsiirola) [inline pyomo/core/base/range.py:615]: This needs an additional `-` in the header line below

File: pyomo/mpec/complementarity.py [hunk 2]
```diff
@@ -45,7 +45,7 @@ def complements(a, b):
 
 class _ComplementarityData(_BlockData):
     def _canonical_expression(self, e):
-        # Note: as the complimentarity component maintains references to
+        # Note: as the complimentary component maintains references to
```
Comment (jsiirola) [inline pyomo/mpec/complementarity.py:48]: This should be `Complimentarity`

File: pyomo/solvers/plugins/solvers/direct_or_persistent_solver.py [hunk 3]
```diff
@@ -67,7 +67,7 @@ def __init__(self, **kwds):
         """A dictionary mapping pyomo constraints to solver constraints."""
 
         self._vars_referenced_by_con = ComponentMap()
-        """A dictionary mapping constraints to a ComponentSet containt the pyomo variables referenced by that
+        """A dictionary mapping constraints to a ComponentSet constaint the pyomo variables referenced by that
```
Comment (jsiirola) [inline pyomo/solvers/plugins/solvers/direct_or_persistent_solver.py:70]: I believe `constaint` should have been `containing`

**Review by jsiirola — CHANGES_REQUESTED** [review]
Just a couple minor edits

**Review by jsiirola — APPROVED** [review]

**Review by Robbybp — APPROVED** [review]
I went through the contrib files and noted a few places where user code *could* break. I'm still in favor of merging, just making sure people are aware.

**Pull request merged**